### ISSUE TITLE
[Bug] cutedsl_bf16_gemm 2-CTA TGV kernel: missing trailing cluster barrier → CUDBG_EXCEPTION_CLUSTER_BLOCK_NOT_PRESENT (SM103, PDL + CUDA graph)

### ISSUE THREAD

**Issue by junliu-mde** [sgl-project/sglang#32907]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] This is a bug report, not a general question.
- [x] Please use English. (A Chinese translation is included in a collapsed section at the end for convenience.)

### Describe the bug

While serving Kimi-K3 (TP8, PD-disaggregated decode, B300/SM103, CUDA graph + PDL), the decode process crashed with `CUDBG_EXCEPTION_CLUSTER_BLOCK_NOT_PRESENT`.

The CUDA core dump shows:

- Faulting kernel: `TgvGemmCuteExtKernel_cta64x16x128_2cta1_pdl1_bias0` (tactic 18: 2-CTA cluster, 12-stage, PDL), on the shared-expert down projection `[8,768] @ [7168,768]^T`.
- The cluster contains only block 1; block 0 has already exited.
- Block 1 is still issuing barrier/TMA events into block 0's SMEM.

In `python/sglang/kernels/ops/gemm/cutedsl_bf16_gemm.py`, cluster synchronization exists only at kernel entry (L378 `cluster_arrive_relaxed`, L438 `cluster_wait`) — there is no cluster barrier before CTA exit, while cross-CTA traffic runs until the end of the kernel (peer-redirected arrives L546/L617, multicast `tcgen05.commit` L747/L754, cluster-coherent TMEM dealloc L763). So one CTA can retire while its peer still has in-flight operations targeting it, which matches the core dump.

The full CUDA core dump (~330 MB) is available on request.

**Proposed fix:** end the kernel body with a cluster-wide barrier for `use_2cta` (all warps):

```python
if cutlass.const_expr(self.use_2cta):
    cute.arch.cluster_arrive()
    cute.arch.cluster_wait()
```

### Reproduction

Production trigger: serving Kimi-K3 (TP8, PD-disaggregated, image `lmsysorg/sglang:kimi-k3@sha256:2e8ef3746b2591287db0f37b4470910ab08974bdaf96fc5820bac35f6d3962bc`); decode crashes intermittently after hours under load.

<details>
<summary>Minimal script + run command + reproduction caveat</summary>

Minimal single-GPU script pinning the faulting configuration (tactic 18, PDL, CUDA graph chain, no weights needed):

```python
#!/usr/bin/env python3

import argparse

import torch

from sglang.kernels.ops.gemm.cutedsl_bf16_gemm import _run_tgv


M = 8
N = 7168
K = 768
TACTIC = 18  # cta64x16x128, 12 stages, 2-CTA cluster


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--iterations", type=int, default=100_000)
    parser.add_argument("--chain-length", type=int, default=8)
    parser.add_argument("--sync-every", type=int, default=100)
    args = parser.parse_args()

    if torch.cuda.get_device_capability() not in ((10, 0), (10, 3)):
        raise RuntimeError("This reproducer requires an SM100 or SM103 GPU")

    device = torch.device("cuda")
    x = torch.empty((M, K), dtype=torch.bfloat16, device=device)
    weight = torch.empty((N, K), dtype=torch.bfloat16, device=device)
    outputs = [
        torch.empty((M, N), dtype=torch.bfloat16, device=device)
        for _ in range(args.chain_length)
    ]

    # Compile tactic 18 before capture. Each launch uses PDL, matching the
    # failing TGV launch mode while keeping the reproducer single-GPU.
    for output in outputs:
        _run_tgv(x, weight.t(), None, output, pdl=True, tactic=TACTIC)
    torch.cuda.synchronize()

    graph = torch.cuda.CUDAGraph()
    with torch.cuda.graph(graph):
        for output in outputs:
            _run_tgv(x, weight.t(), None, output, pdl=True, tactic=TACTIC)

    print(
        f"device={torch.cuda.get_device_name()} shape=({M},{K})x({N},{K}) "
        f"tactic={TACTIC} chain_length={args.chain_length}"
    )
    for iteration in range(1, args.iterations + 1):
        graph.replay()
        if iteration % args.sync_every == 0:
            torch.cuda.synchronize()
            print(f"completed={iteration}", flush=True)

    torch.cuda.synchronize()


if __name__ == "__main__":
    main()
```

Run with device-side core dumps enabled to capture the faulting kernel on crash:

```bash
CUDA_ENABLE_COREDUMP_ON_EXCEPTION=1 CUDA_COREDUMP_FILE=/tmp/core.cuda.%h.%p \
python3 kimi_k3_tgv_cluster_repro.py
```

Honest caveat: this synthetic script did **not** trigger the exception in ~2.5e8 kernel launches on idle GPUs, including runs with concurrent matmul contention streams. With symmetric per-CTA load the two CTAs stay in lockstep, so the race window never opens; the production crash needed real scheduling skew (the PDL successor kernel stealing SMs mid-graph). The root-cause claim rests on the CUDA core dump analysis above (block 0 exited, block 1 still emitting DSM traffic toward it), not on this script. The script documents the exact faulting kernel configuration.

</details>

### Environment

```
Python: 3.12.3
GPU 0-7: NVIDIA B300 SXM6 AC (Compute Capability 10.3)
NVCC: CUDA 13.0, V13.0.88
CUDA Driver Version: 580.126.20
PyTorch: 2.11.0+cu130
sglang: 0.5.16
sglang-kernel: 0.4.5
flashinfer_python: 0.6.15.post1
triton: 3.6.0
```

<details>
<summary>中文版本 (Chinese version)</summary>

### 问题描述

Kimi-K3 服务（TP8，PD 分离 decode，B300/SM103，CUDA graph + PDL）的 decode 进程崩溃，异常为 `CUDBG_EXCEPTION_CLUSTER_BLOCK_NOT_PRESENT`。

CUDA core dump 显示：

- 故障 kernel：`TgvGemmCuteExtKernel_cta64x16x128_2cta1_pdl1_bias0`（tactic 18：2-CTA cluster、12 stage、PDL），对应 shared expert down projection `[8,768] @ [7168,768]^T`。
- cluster 中只剩 block 1，block 0 已退出。
- block 1 仍在向 block 0 的 SMEM 发 barrier/TMA 事件。

`python/sglang/kernels/ops/gemm/cutedsl_bf16_gemm.py` 中，cluster 同步只在 kernel 入口有（L378、L438）；CTA 退出前没有 cluster barrier，而跨 CTA 通信一直持续到 kernel 结尾（L546/L617 重定向 arrive、L747/L754 多播 commit、L763 cluster 级 TMEM dealloc）。因此一个 CTA 可能在对端仍有指向它的未完成操作时退出，与 core dump 现场一致。

完整 core dump（约 330 MB）可按需提供。

**建议修复**：kernel 结尾对 `use_2cta` 增加 cluster 级 barrier（所有 warp 执行）：

```python
if cutlass.const_expr(self.use_2cta):
    cute.arch.cluster_arrive()
    cute.arch.cluster_wait()
```

### 复现说明

如实说明：上面的合成脚本在空闲 GPU 上约 2.5 亿次 kernel 发射（含并发 matmul 争抢流）未触发该异常。两个 CTA 负载对称时进度锁死，竞态窗口打不开；生产崩溃需要真实的调度偏差（PDL 后继 kernel 在 graph 中途抢占 SM）。根因结论基于上述 core dump 分析（block 0 已退出、block 1 仍向其发 DSM 事件），不依赖该脚本；脚本的作用是固定故障 kernel 的准确配置。

</details>

**Comment by junliu-mde** [comment]
**Update: 8 more occurrences in one day, on-demand reproduction, and a patch under live validation**

Since filing, this fired 8 more times within one day on the same B300 (SM103) PD deployment (decode side, TP8 + DCP8, PDL + CUDA graph). I verified the two most recent GPU coredumps with `cuda-gdb` — identical signature, byte-identical faulting PC:

```
CUDA Exception: Cluster target block not present
#0  0x00000091554787e0 in kernel_cutlass_kernel_TgvGemmCuteExtKernel_cta64x16x128_2cta1_pdl1_bias0_..._0<<<(112,1,1),(256,1,1)>>>
concurrently resident: sglang::all_reduce_pull_res_kernel<8u, false, true><<<(4,1,1),(512,1,1)>>>

occurrence A: faulting rank 5, cluster (0,0,0):  surviving block (1,0,0)   -> exited peer blockIdx(0,0,0)
occurrence B: faulting rank 7, cluster (51,0,0): surviving block (103,0,0) -> exited peer blockIdx(102,0,0)
```

Only the odd CTA of the 2-CTA cluster is still resident; the faulting TP rank and cluster coordinate differ per occurrence — a lifetime race, not a data-dependent fault.

Two observations that may help reproduction elsewhere:

- A decode worker that had served batch 1–6 for ~27 min crashed **within ~1 s** of 12 concurrent long-context streams (20k–150k input tokens, 1–4k output, `ignore_eos`) arriving at the router. Those requests were still in tokenize/prefill at fault time, so the host-side scheduling perturbation alone (a burst of preallocations shifting launch timing under PDL) appears to be enough to open the race window. This matches the earlier finding that ~2.5e8 fixed-shape synthetic launches on an idle GPU never trigger it — you need launch-timing jitter, not launch count.
- The non-faulting ranks show a second flavor of the same bug: they wedge in `torch.cuda.Stream.synchronize()` inside `process_batch_result_decode` until the scheduler watchdog aborts. Without coredump-on-exception enabled, this bug can present as a "GPU spin hang" rather than a crash.

**Patch under test** — functionally identical to the current HEAD of #32954 (relaxed arrive + wait at the warp-dispatch join). The comment block additionally records why the arrive cannot usefully be hoisted earlier, which was the one open review question there:

```diff
--- a/python/sglang/kernels/ops/gemm/cutedsl_bf16_gemm.py
+++ b/python/sglang/kernels/ops/gemm/cutedsl_bf16_gemm.py
@@ -489,6 +489,21 @@
                 c_dtype,
                 d_layout,
             )

+        if cutlass.const_expr(self.use_2cta):
+            cute.arch.cluster_arrive_relaxed()
+            cute.arch.cluster_wait()
+
     # ====================================================================
     # DMA_A WARP — TMA-loads A tiles into sA[..., stage], one per K-iter.
```

- `const_expr(self.use_2cta)` keeps 1-CTA tactics byte-identical.
- The insertion point is the warp-dispatch join reached by all 256 threads (warp 3 falls through immediately; no early returns inside the branch functions), satisfying the aligned cluster-arrive requirement. PDL successor kernels stealing SMs can delay the arrive but not prevent it, so no deadlock path.

**Live validation in progress:** patched decode workers are soaking under the same 12-stream reproduction load plus background traffic — conditions that previously killed a worker within seconds to minutes. Zero new exceptions or coredumps so far; I will follow up with soak results.

<details>
<summary>中文版</summary>

**进展**:提交 issue 后一天内同一 B300 (SM103) PD 部署的 decode 侧又崩 8 次。用 `cuda-gdb` 核对最近两个 GPU coredump:同一 kernel、同一故障指令地址 `0x91554787e0`、同样只剩 2-CTA cluster 的奇数块、同样有 all-reduce 并发。故障 rank 和 cluster 坐标每次不同(rank5/cluster0 与 rank7/cluster51)——生命期竞态特征。

**复现要点**:某 decode 以 batch 1–6 稳定运行约 27 分钟,12 路长上下文并发流(输入 20k–150k token、输出 1–4k、`ignore_eos`)到达 router 后约 1 秒内即触发设备异常;当时新请求还在 tokenize/prefill,说明 host 侧发射时序扰动本身就能打开竞态窗口。这与"空闲 GPU 上 2.5 亿次固定形状发射不触发"一致:需要的是时序抖动,不是次数。另外,非故障 rank 会卡死在 `stream.synchronize` 直到 watchdog 中止——不开 coredump-on-exception 时,这个 bug 会表现为"GPU 自旋 hang"而非崩溃。

**在测 patch**(与 #32954 当前 HEAD 功能一致:warp dispatch 汇合点加 relaxed arrive + wait;注释额外论证了 arrive 无法更早):见上方 diff。CuTeDSL JIT kernel,无需重编 `sgl-kernel`,覆盖 `.py` 重启即生效;`const_expr(use_2cta)` 保证 1-CTA 路径零改动。

**验证中**:打了 patch 的 decode 正在同一套复现负载+背景流量下压测(此前该条件下几秒到几分钟必崩),目前零新异常/零新 coredump,后续跟进 soak 结果。

</details>
